crates/ra_parser/src/grammar/items.rs

```diff
@@ -35,54 +35,42 @@ pub(super) const ITEM_RECOVERY_SET: TokenSet = token_set![
 ];
 
 pub(super) fn item_or_macro(p: &mut Parser, stop_on_r_curly: bool, flavor: ItemFlavor) {
-    let m = p.start();
+    let mut m = p.start();
     attributes::outer_attributes(p);
-    match maybe_item(p, flavor) {
-        MaybeItem::Item(kind) => {
-            m.complete(p, kind);
-        }
-        MaybeItem::None => {
-            if paths::is_path_start(p) {
-                match macro_call(p) {
-                    BlockLike::Block => (),
-                    BlockLike::NotBlock => {
-                        p.expect(SEMI);
-                    }
-                }
-                m.complete(p, MACRO_CALL);
-            } else {
-                m.abandon(p);
-                if p.at(L_CURLY) {
-                    error_block(p, "expected an item");
-                } else if p.at(R_CURLY) && !stop_on_r_curly {
-                    let e = p.start();
-                    p.error("unmatched `}`");
-                    p.bump();
-                    e.complete(p, ERROR);
-                } else if !p.at(EOF) && !p.at(R_CURLY) {
-                    p.err_and_bump("expected an item");
-                } else {
-                    p.error("expected an item");
-                }
+    m = match maybe_item(p, m, flavor) {
+        Some(m) => m,
+        None => return,
```

Comment:
This gave me a pause: `Option` has an inverted modality here, where `None` means "success" and `Some` is an error. What do you think about making the return type be `Result<(), Marker>` instead?